### PR TITLE
Post-merge `snakemake.params`

### PR DESCRIPTION
**Pull request by virio-andreyana**
# This is a re-do for the pull request in #296.

## Changes proposed in this Pull Request

Use the `params:` section in rule definition to keep track of changed settings in `config.yaml`.

https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#non-file-parameters-for-rules

The goal is that rules for which parameters have changed rerun automatically.

This can currently be done with:

https://snakemake.readthedocs.io/en/stable/project_info/faq.html#how-do-i-trigger-re-runs-for-rules-with-updated-code-or-parameters

```sh
snakemake -n -R `snakemake --list-params-changes`
```

or

```sh
snakemake -R $(snakemake --list-params-changes)
```

This will be even more convenient, once this `snakemake` PR is released:

https://github.com/snakemake/snakemake/pull/1107

```sh
snakemake --rerun-params-changed
```

or 

```sh
snakemake --rerun-changed
```

(which would also include code changes)

It seems to work very nicely to pass lists/dicts as params, which would help us keep config "blocks".

This PR is not complete yet, but a proof-of-concept for a selection of config settings. We can discuss from here and complete later.

## Rules Checklist

- [x] Snakefile
- [x] build_electricity.smk
- [x] build_sector.smk
- [x] collect.smk
- [x] common.smk
- [x] postprocess.smk
- [x] retrieve.smk
- [x] solve_electricity.smk
- [x] solve_myopic.smk
- [x] solve overnight.smk

## Script Checklist

- [ ] _helpers.py 
- [x] add_brownfield.py
- [x] add_electricity.py
- [x] add_existing_baseyear.py
- [x] add_extra_components.py
- [x] base_network.py
- [x] build_ammonia_production.py
- [x] build_biomass_potentials.py
- [x] build_biomass_transport_costs.py
- [x] build_bus_regions.py
- [x] build_clustered_population_layouts.py
- [x] build_cop_profiles.py
- [x] build_cutout.py
- [x] build_electricity_demand.py
- [x] build_energy_totals.py
- [x] build_gas_input_locations.py
- [x] build_gas_network.py
- [x] build_heat_demand.py
- [x] build_hydro_profile.py
- [x] build_industrial_distribution_key.py
- [x] build_industrial_energy_demand_per_country_today.py
- [x] build_industrial_energy_demand_per_node_today.py
- [x] build_industrial_energy_demand_per_node.py
- [x] build_industrial_production_per_country_tomorrow.py
- [x] build_industrial_production_per_country.py
- [x] build_industrial_production_per_node.py
- [x] build_industry_sector_ratios.py
- [x] build_natura_raster.py
- [x] build_population_layouts.py
- [x] build_population_weighted_energy_totals.py
- [x] build_powerplants.py
- [x] build_renewable_profiles.py
- [x] build_retro_cost.py
- [x] build_salt_cavern_potentials.py
- [x] build_sequestration_potentials.py
- [x] build_shapes.py
- [x] build_ship_raster.py
- [x] build_shipping_demand.py
- [x] build_solar_thermal_profiles.py
- [x] build_temperature_profiles.py
- [x] build_transport_demand.py
- [x] cluster_gas_network.py
- [x] cluster_network.py
- [x] copy_config.py
- [x] make_summary.py
- [x] plot_network.py
- [x] plot_summary.py
- [x] prepare_links_p_nom.py
- [x] prepare_network.py
- [x] prepare_sector_network.py
- [x] retrieve_databundle.py
- [x] retrieve_gas_infrastructure_data.py
- [x] retrieve_sector_databundle.py
- [x] simplify_network.py
- [x] solve_network.py
- [x] solve_operations_network.py

## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [ ] Code and workflow changes are sufficiently documented.
- [x] Changed dependencies are added to `envs/environment.yaml`.
- [x] Changes in configuration options are added in all of `config.default.yaml`.
- [ ] Changes in configuration options are also documented in `doc/configtables/*.csv`.
- [ ] A release note `doc/release_notes.rst` is added.

### CODE REVIEW

**Comment by virio-andreyana** [comment]
The majority of the param conversion is done by a python algorithm. Some changes are simple, only changing the word "config" into "param". The complicated part is in `simplify_network.py` and `solve_network.py`. While the `simplify_network.py` has been amended, `solve_network.py` is the hardest. It ask for almost every variable variable there is in config. 

I believe its better to just add in param `config=config` since any changes that happens in config require us to rerun the solve_network.py anyway. But what do you think?

**Comment by FabianHofmann** [comment]
Hey @virio-andreyana, I would not pass the whole config but use the sections which are really needed (even if it is a lot). Imagine you just want to alter some colours for the visualization of results (which happens after the solving), then the whole solving process would be retriggered.

File: rules/build_electricity.smk [hunk 1]
```diff
@@ -235,6 +252,10 @@ rule build_renewable_profiles:
 
 
 rule build_hydro_profile:
+    params:
+        hydro=config["renewable"]["hydro"],
+        countries=config["countries"],
+        renewable=config["renewable"],
```
Comment (FabianHofmann) [inline rules/build_electricity.smk:258]: we can remove that

File: rules/build_sector.smk [hunk 2]
```diff
@@ -386,6 +404,42 @@ rule build_ammonia_production:
 
 
 rule build_industry_sector_ratios:
+    params:
+        industry={
```
Comment (FabianHofmann) [inline rules/build_sector.smk:408]: use higher key

File: rules/build_sector.smk [hunk 3]
```diff
@@ -426,6 +492,19 @@ rule build_industrial_production_per_country:
 
 
 rule build_industrial_production_per_country_tomorrow:
+    params:
+        industry={
```
Comment (FabianHofmann) [inline rules/build_sector.smk:496]: use higher key

File: rules/build_sector.smk [hunk 4]
```diff
@@ -524,6 +606,14 @@ rule build_industrial_energy_demand_per_node:
 
 
 rule build_industrial_energy_demand_per_country_today:
+    params:
+        countries=config["countries"],
+        industry={
```
Comment (FabianHofmann) [inline rules/build_sector.smk:611]: use higher key

File: rules/build_sector.smk [hunk 5]
```diff
@@ -666,6 +762,30 @@ rule build_transport_demand:
 
 rule prepare_sector_network:
     params:
+        co2_budget=config["co2_budget"],
+        solver_name=config["solving"]["solver"]["name"],
+        conventional_carriers=config["existing_capacities"]["conventional_carriers"],
+        foresight=config["foresight"],
+        costs=config["costs"],
+        sector=config["sector"],
+        industry={
```
Comment (FabianHofmann) [inline rules/build_sector.smk:771]: use higher key

File: rules/solve_electricity.smk [hunk 6]
```diff
@@ -4,6 +4,18 @@
 
 
 rule solve_network:
+    params:
+        solving=config["solving"],
+        config_parts={
```
Comment (FabianHofmann) [inline rules/solve_electricity.smk:9]: can we define the dictionary in the script instead? and use direct assignments here instead?

File: rules/solve_myopic.smk [hunk 7]
```diff
@@ -74,6 +83,18 @@ ruleorder: add_existing_baseyear > add_brownfield
 
 
 rule solve_sector_network_myopic:
+    params:
+        solving=config["solving"],
+        config_parts={
```
Comment (FabianHofmann) [inline rules/solve_myopic.smk:88]: same here

File: rules/solve_overnight.smk [hunk 8]
```diff
@@ -4,6 +4,18 @@
 
 
 rule solve_sector_network:
+    params:
+        solving=config["solving"],
+        config_parts={
```
Comment (FabianHofmann) [inline rules/solve_overnight.smk:9]: same here

**Review by FabianHofmann — CHANGES_REQUESTED** [review]

**Review by FabianHofmann — APPROVED** [review]

**Pull request merged**